cli/src/commands/metadata.rs

```diff
@@ -25,27 +25,48 @@ pub struct Opts {
     /// when using the option.
     #[clap(long, use_value_delimiter = true, value_parser)]
     pallets: Option<Vec<String>>,
+    /// Generate a subset of the metadata that contains only the
+    /// runtime APIs needed.
+    ///
+    /// The returned metadata is updated to the latest available version
+    /// when using the option.
+    #[clap(long, use_value_delimiter = true, value_parser)]
+    runtime_apis: Option<Vec<String>>,
 }
 
 pub async fn run(opts: Opts) -> color_eyre::Result<()> {
     let bytes = opts.file_or_url.fetch().await?;
     let mut metadata = <RuntimeMetadataPrefixed as Decode>::decode(&mut &bytes[..])?;
 
-    if let Some(pallets) = opts.pallets {
+    if opts.pallets.is_some() || opts.runtime_apis.is_some() {
         let mut metadata_v15 = match metadata.1 {
             RuntimeMetadata::V14(metadata_v14) => metadata_v14_to_latest(metadata_v14),
             RuntimeMetadata::V15(metadata_v15) => metadata_v15,
             _ => {
                 return Err(eyre::eyre!(
                     "Unsupported metadata version {:?}, expected V14.",
                     metadata.1
-                ))
+                ));
             }
         };
-
-        retain_metadata_pallets(&mut metadata_v15, |pallet_name| {
-            pallets.iter().any(|p| &**p == pallet_name)
-        });
+        match (opts.pallets.as_ref(), opts.runtime_apis.as_ref()) {
+            (Some(pallets), Some(runtime_apis)) => retain_metadata(
+                &mut metadata_v15,
+                |pallet_name| pallets.iter().any(|p| &**p == pallet_name),
+                |runtime_api_name| runtime_apis.iter().any(|p| &**p == runtime_api_name),
+            ),
+            (Some(pallets), None) => retain_metadata(
+                &mut metadata_v15,
+                |pallet_name| pallets.iter().any(|p| &**p == pallet_name),
+                |_| true,
+            ),
+            (None, Some(runtime_apis)) => retain_metadata(
+                &mut metadata_v15,
+                |_| true,
+                |runtime_api_name| runtime_apis.iter().any(|p| &**p == runtime_api_name),
+            ),
+            (None, None) => {}
+        }
         metadata = metadata_v15.into();
```

Comment:
The way to avoid the repetition would be by boxing the closures like so:

```rust
let retain_pallets_fn: Box<dyn Fn(&str) -> bool> = match opts.pallets.as_ref() {
    Some(pallets) => Box::new(|name| pallets.iter().any(|p| &**p == name)),
    None => Box::new(|_| true)
};
let retain_runtime_apis_fn: Box<dyn Fn(&str) -> bool> = match opts.runtime_apis.as_ref() {
    Some(apis) => Box::new(|name| apis.iter().any(|p| &**p == name)),
    None => Box::new(|_| true)
};
retain_metadata(&mut metadata_v15, retain_pallets_fn, retain_runtime_apis_fn);
metadata = metadata_v15.into();
```

Reply:
Ah, nice, yeah that is definitely a solution. I still want to keep a check that at least one of the arguments is set, because if not we can avoid calling the `retain_metadata` function alltogether.

Reply:
Yeah I wouldn't reaplce the initial check, just the highlighted lines :)